### PR TITLE
CMake with FF7Tk

### PR DESCRIPTION
**Pull request by myst6re**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,42 +17,50 @@
...
+
+# specify the C++ standard
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+
+set(CMAKE_AUTOMOC ON)
+set(CMAKE_AUTORCC ON)
+set(CMAKE_AUTOUIC ON)
+
+if(CMAKE_VERSION VERSION_LESS "3.7.0")
```
Comment (sithlord48) [inline CMakeLists.txt:32]: Impossible condition you require cmake 3.15

File: CMakeLists.txt [hunk 2]
```diff
@@ -17,42 +17,50 @@
...
+
+# Translations
+file(GLOB_RECURSE translation_files "${RELEASE_NAME}_*.ts")
+set_source_files_properties(${translation_files} PROPERTIES OUTPUT_LOCATION "${CMAKE_INSTALL_FULL_BINDIR}")
+qt5_add_translation(QM_FILES ${translation_files})
+
+if(WIN32)
+    set (FF7TK_DEPENDS
+        ${ff7tk_DIR}/../../../bin/ff7tk.dll
+        ${ff7tk_DIR}/../../../bin/ff7tkWidgets.dll
```
Comment (sithlord48) [inline CMakeLists.txt:54]: are you using widgets or utils?

**Review by sithlord48 — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (myst6re) [reply CMakeLists.txt:54]: Utils, this line shouldn't be here yet, as FF7TK_DEPENDS is used for install rules.

File: CMakeLists.txt [hunk 3]
```diff
@@ -19,7 +19,11 @@
 cmake_minimum_required(VERSION 3.15)
 cmake_policy(SET CMP0074 NEW)
 
-project(Makou_Reactor LANGUAGES CXX)
+set(RELEASE_NAME "Makou_Reactor")
```
Comment (sithlord48) [inline CMakeLists.txt:22]: the project call sets your Name and Release Version 

`project(makoureactor VERSION 1.8.5 DESCRIPTION Edit all your FF7 Fields LANGUAGES CXX)`

File: CMakeLists.txt [hunk 4]
```diff
@@ -29,38 +33,54 @@ set(CMAKE_AUTOMOC ON)
 set(CMAKE_AUTORCC ON)
```
Comment (sithlord48) [inline CMakeLists.txt:33]: You should also `set(CMAKE_AUTOMOC ON)`

File: CMakeLists.txt [hunk 5]
```diff
@@ -29,38 +33,54 @@ set(CMAKE_AUTOMOC ON)
...
-set(GUI_TARGET "${RELEASE_NAME}")
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
 
-find_package(Qt5 COMPONENTS Widgets OpenGL LinguistTools REQUIRED)
+find_package(QT NAMES Qt5 COMPONENTS Widgets OpenGL LinguistTools REQUIRED)
+find_package(Qt${QT_VERSION_MAJOR} COMPONENTS Widgets OpenGL LinguistTools REQUIRED)
 find_package(ff7tk REQUIRED NO_MODULE COMPONENTS ff7tk ff7tkUtils)
 find_package(ZLIB REQUIRED)
 
+file(GLOB_RECURSE TS_FILES "${RELEASE_NAME}_*.ts")
```
Comment (sithlord48) [inline CMakeLists.txt:43]: Generate them in their own sub dir as seen here: https://github.com/sithlord48/ff7tk/tree/master/lang

File: CMakeLists.txt [hunk 6]
```diff
@@ -29,38 +33,54 @@ set(CMAKE_AUTOMOC ON)
...
+set(RESOURCES "src/qt/${RELEASE_NAME}.qrc")
 
 # GUI
-add_executable(${GUI_TARGET} ${PROJECT_SOURCES} ${QM_FILES} src/qt/${RELEASE_NAME}.qrc src/qt/${RELEASE_NAME}.rc)
+add_executable(${GUI_TARGET} ${PROJECT_SOURCES} ${QM_FILES} ${RESOURCES} src/qt/${RELEASE_NAME}.rc)
 target_include_directories(${GUI_TARGET} PRIVATE "${CMAKE_SOURCE_DIR}/src")
 target_link_libraries(${GUI_TARGET}
-    Qt5::OpenGL
-    Qt5::Widgets
+    Qt${QT_VERSION_MAJOR}::OpenGL
```
Comment (sithlord48) [inline CMakeLists.txt:51]: If your Building w/ Qt 5.15 + you can just use `Qt::OpenGL` ...

File: CMakeLists.txt [hunk 7]
```diff
@@ -29,38 +33,54 @@ set(CMAKE_AUTOMOC ON)
...
-    set (FF7TK_DEPENDS
-        ${ff7tk_DIR}/../../../bin/ff7tk.dll
-        ${ff7tk_DIR}/../../../bin/ff7tkWidgets.dll
-    )
-endif()
+set(RESOURCES "src/qt/${RELEASE_NAME}.qrc")
 
 # GUI
-add_executable(${GUI_TARGET} ${PROJECT_SOURCES} ${QM_FILES} src/qt/${RELEASE_NAME}.qrc src/qt/${RELEASE_NAME}.rc)
+add_executable(${GUI_TARGET} ${PROJECT_SOURCES} ${QM_FILES} ${RESOURCES} src/qt/${RELEASE_NAME}.rc)
```
Comment (sithlord48) [inline CMakeLists.txt:48]: You should add Both WIN32 and MACOSX_BUNDLE between ${GUI_TARGET} ${PROJECT_SOURCES}
Tldr this makes sure on win32 you build a gui application and on mac os you create a proper bundle (i.e .app file)

File: CMakeLists.txt [hunk 8]
```diff
@@ -29,38 +33,54 @@ set(CMAKE_AUTOMOC ON)
...
+#target_link_libraries(${CLI_TARGET}
+#    Qt${QT_VERSION_MAJOR}::Widgets
+#    ff7tk::ff7tk
+#    ff7tk::ff7tkUtils
+#)
+#target_compile_definitions(${CLI_TARGET}
+#    PRIVATE MR_CONSOLE=1
+#)
+
+qt5_create_translation(QM_FILES ${CMAKE_SOURCE_DIR} ${TS_FILES})
```
Comment (sithlord48) [inline CMakeLists.txt:70]: Becareful here without making them in their own directory you may find on clean your TS files are deleted.
see : https://github.com/sithlord48/ff7tk/tree/master/lang

Comment (sithlord48) [reply CMakeLists.txt:70]: qt5_create_translations does not work on qt6 .. you should use qt_ in place of any qt5_ or qt6_ calls if you plan to allow buildng with both versions

**Review by sithlord48 — CHANGES_REQUESTED** [review]

**Review by sithlord48 — CHANGES_REQUESTED** [review]

File: lang/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with this program.  If not, see <http://www.gnu.org/licenses/>.
+###############################################################################
+
+find_package(QT NAMES Qt5 COMPONENTS LinguistTools)
+find_package(Qt${QT_VERSION_MAJOR}LinguistTools)
+
+file(GLOB_RECURSE TS_FILES "${RELEASE_NAME}_*.ts")
```
[the remark this answers is not in the document]
Comment (myst6re) [reply lang/CMakeLists.txt:22]: Using file with GLOB_RECURSE to set TS_FILES is not the same thing?

File: CMakeLists.txt [hunk 10]
```diff
@@ -17,42 +17,70 @@
...
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+
+set(CMAKE_AUTOMOC ON)
+set(CMAKE_AUTORCC ON)
+set(CMAKE_AUTOUIC ON)
+
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+find_package(QT NAMES Qt5 COMPONENTS Widgets OpenGL REQUIRED)
```
[the remark this answers is not in the document]
Comment (myst6re) [reply CMakeLists.txt:38]: About that, is ff7tk compatible with Qt 6? Last time I checked it didn't compile

File: lang/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+qt5_create_translation(QM_FILES ${CMAKE_SOURCE_DIR} ${TS_FILES})
```
[the remark this answers is not in the document]
Comment (myst6re) [reply lang/CMakeLists.txt:24]: I haven't done the install part yet. What's the target? Is it a custom command like `cmake translations`?

File: lang/CMakeLists.txt (hunk 9 above)
[the remark this answers is not in the document]
Comment (sithlord48) [reply lang/CMakeLists.txt:24]: No qt_create_translation makes the ts files if they do not exist or updates existing ones. If you have a GLOB list here you can never have it generate a new one with out first making it manually.

File: lang/CMakeLists.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (sithlord48) [reply lang/CMakeLists.txt]: This is basicly saying . we have a new target named
Translations it required by ALL other targets and for this to be complete it DEPENDS on the ${QM_FILES}
So that any build of anything will trigger this. 

Then later when we do an install this will install the following files. install rules are created if the target is  built

File: lang/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+###############################################################################
+
+set_directory_properties(PROPERTIES CLEAN_NO_CUSTOM 1)
+
+find_package(QT NAMES Qt5 Qt6 COMPONENTS LinguistTools)
+find_package(Qt${QT_VERSION_MAJOR}LinguistTools)
+
+set(TS_FILES
+    Makou_Reactor_fr.ts
+    Makou_Reactor_ja.ts
```
Comment (sithlord48) [inline lang/CMakeLists.txt:26]: You should add also english  for things like plurals.

File: CMakeLists.txt [hunk 13]
```diff
@@ -17,42 +17,109 @@
...
+
+find_package(QT NAMES Qt5 Qt6 COMPONENTS Widgets OpenGL Zlib REQUIRED)
+find_package(Qt${QT_VERSION_MAJOR} COMPONENTS Widgets OpenGL Zlib REQUIRED)
+find_package(ff7tk COMPONENTS ff7tk ff7tkUtils REQUIRED)
+
+file(GLOB_RECURSE PROJECT_SOURCES "src/*.cpp" "src/*.h")
+set(RESOURCES "src/qt/${RELEASE_NAME}.qrc")
+
+if (APPLE)
+    set(OS_BUNDLE MACOSX_BUNDLE)
```
Comment (sithlord48) [inline CMakeLists.txt:48]: Both WIN32 and MACOSX_BUNDLE can be  in the same add_executable. They only have an effect on their platforms and no others.
